include/boost/beast/http/impl/read.ipp

```diff
@@ -34,17 +35,30 @@ namespace detail {
 
 //------------------------------------------------------------------------------
 
+template<typename DynamicBuffer>
+boost::optional<typename DynamicBuffer::mutable_buffers_type>
+prepare_or_error(DynamicBuffer& buffer, error_code& ec)
+{
+    try
+    {
+        ec.assign(0, ec.category());
```

Comment:
If there's an error we are paying for clearing `ec` needlessly. It might be would be more efficient to clear the error if the call to `prepare` succeeds. We would have to study the output in Compiler Explorer (https://godbolt.org/)

Reply:
I'll check it out, though it's worth remembering the the most likely case is that there is no error. And if there is one, we've wasted 2 machine word writes, so I don't think this is a huge issue.

Reply:
True

Reply:
I've checked it and the difference is 1 assignment. One thing I'm wondering about is how should a composed operation behave if prepare() throws a different error (e.g. bad_alloc)?

Reply:
> how should a composed operation behave if prepare() throws a different error (e.g. bad_alloc)

I think it should propagate the exception as normal.